File: public/sdk/CulturalTranslateSDK.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using UnityEngine;
using UnityEngine.Networking;

namespace CulturalTranslate
{
    /// <summary>
    /// Cultural Translate SDK for Unity
    /// Real-time translation for multiplayer games
    /// </summary>
    public class CulturalTranslateSDK : MonoBehaviour
    {
        [Header("Configuration")]
        public string apiKey;
        public string gameId;
        public string baseUrl = "https://culturaltranslate.com/api";
        
        [Header("Settings")]
        public string sourceLanguage = "en";
        public string targetLanguage = "ar";
        public string sourceCulture = "en-US";
        public string targetCulture = "ar-SA";
        public string culturalAdaptationLevel = "standard";
        
        // Events
        public event Action<TranslationResult> OnTranslation;
        public event Action<string> OnError;
        public event Action OnConnected;
        public event Action OnDisconnected;
        
        private string sessionId;
        private bool isInitialized = false;
        
        void Start()
        {
            if (string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(gameId))
            {
                Debug.LogError("[CulturalTranslate] API Key and Game ID are required!");
                return;
            }
            
            Debug.Log("[CulturalTranslate] SDK Initialized");
            isInitialized = true;
        }
        
        /// <summary>
        /// Create a new translation session
        /// </summary>
        public void CreateSession(Action<string> onSuccess = null)
        {
            if (!isInitialized)
            {
                OnError?.Invoke("SDK not initialized");
                return;
            }
            
            StartCoroutine(CreateSessionCoroutine(onSuccess));
        }
        
        private IEnumerator CreateSessionCoroutine(Action<string> onSuccess)
        {
            var requestData = new SessionRequest
            {
                source_language = sourceLanguage,
                target_language = targetLanguage,
                source_culture = sourceCulture,
                target_culture = targetCulture,
                type = "gaming",
                cultural_adaptation_level = culturalAdaptationLevel
            };
            
            string jsonData = JsonUtility.ToJson(requestData);
            
            using (UnityWebRequest request = new UnityWebRequest($"{baseUrl}/gaming/sessions", "POST"))
            {
                byte[] bodyRaw = Encoding.UTF8.GetBytes(jsonData);
                request.uploadHandler = new UploadHandlerRaw(bodyRaw);
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Content-Type", "application/json");
                request.SetRequestHeader("Authorization", $"Bearer {apiKey}");
                request.SetRequestHeader("X-Game-ID", gameId);
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    var response = JsonUtility.FromJson<SessionResponse>(request.downloadHandler.text);
                    sessionId = response.public_id;
                    
                    Debug.Log($"[CulturalTranslate] Session created: {sessionId}");
                    OnConnected?.Invoke();
                    onSuccess?.Invoke(sessionId);
                }
                else
                {
                    Debug.LogError($"[CulturalTranslate] Error creating session: {request.error}");
                    OnError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Translate text
        /// </summary>
        public void TranslateText(string text, Action<string> onSuccess = null)
        {
            if (string.IsNullOrEmpty(sessionId))
            {
                OnError?.Invoke("No active session. Create a session first.");
                return;
            }
            
            StartCoroutine(TranslateTextCoroutine(text, onSuccess));
        }
        
        private IEnumerator TranslateTextCoroutine(string text, Action<string> onSuccess)
        {
            var requestData = new TextTranslationRequest
            {
                session_id = sessionId,
                text = text,
                source_language = sourceLanguage,
                target_language = targetLanguage
            };
            
            string jsonData = JsonUtility.ToJson(requestData);
            
            using (UnityWebRequest request = new UnityWebRequest($"{baseUrl}/gaming/translate/text", "POST"))
            {
                byte[] bodyRaw = Encoding.UTF8.GetBytes(jsonData);
                request.uploadHandler = new UploadHandlerRaw(bodyRaw);
                request.downloadHandler = new DownloadHandlerBuffer();
                request.SetRequestHeader("Content-Type", "application/json");
                request.SetRequestHeader("Authorization", $"Bearer {apiKey}");
                request.SetRequestHeader("X-Game-ID", gameId);
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    var response = JsonUtility.FromJson<TranslationResponse>(request.downloadHandler.text);
                    
                    Debug.Log($"[CulturalTranslate] Translation: {text} -> {response.translated_text}");
                    
                    var result = new TranslationResult
                    {
                        sourceText = text,
                        translatedText = response.translated_text,
                        sourceLanguage = sourceLanguage,
                        targetLanguage = targetLanguage
                    };
                    
                    OnTranslation?.Invoke(result);
                    onSuccess?.Invoke(response.translated_text);
                }
                else
                {
                    Debug.LogError($"[CulturalTranslate] Translation error: {request.error}");
                    OnError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// Translate audio (voice)
        /// </summary>
        public void TranslateVoice(byte[] audioData, Action<string> onSuccess = null)
        {
            if (string.IsNullOrEmpty(sessionId))
            {
                OnError?.Invoke("No active session. Create a session first.");
                return;
            }
            
            StartCoroutine(TranslateVoiceCoroutine(audioData, onSuccess));
        }
        
        private IEnumerator TranslateVoiceCoroutine(byte[] audioData, Action<string> onSuccess)
        {
            List<IMultipartFormSection> formData = new List<IMultipartFormSection>();
            formData.Add(new MultipartFormDataSection("session_id", sessionId));
            formData.Add(new MultipartFormFileSection("audio", audioData, "voice.wav", "audio/wav"));
            
            using (UnityWebRequest request = UnityWebRequest.Post($"{baseUrl}/gaming/translate/voice", formData))
            {
                request.SetRequestHeader("Authorization", $"Bearer {apiKey}");
                request.SetRequestHeader("X-Game-ID", gameId);
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    var response = JsonUtility.FromJson<VoiceTranslationResponse>(request.downloadHandler.text);
                    
                    Debug.Log($"[CulturalTranslate] Voice translation completed");
                    
                    var result = new TranslationResult
                    {
                        sourceText = response.source_text,
                        translatedText = response.translated_text,
                        sourceLanguage = sourceLanguage,
                        targetLanguage = targetLanguage,
                        audioUrl = response.audio_url
                    };
                    
                    OnTranslation?.Invoke(result);
                    onSuccess?.Invoke(response.translated_text);
                }
                else
                {
                    Debug.LogError($"[CulturalTranslate] Voice translation error: {request.error}");
                    OnError?.Invoke(request.error);
                }
            }
        }
        
        /// <summary>
        /// End the current session
        /// </summary>
        public void EndSession()
        {
            if (string.IsNullOrEmpty(sessionId))
            {
                return;
            }
            
            StartCoroutine(EndSessionCoroutine());
        }
        
        private IEnumerator EndSessionCoroutine()
        {
            using (UnityWebRequest request = UnityWebRequest.Delete($"{baseUrl}/gaming/sessions/{sessionId}"))
            {
                request.SetRequestHeader("Authorization", $"Bearer {apiKey}");
                request.SetRequestHeader("X-Game-ID", gameId);
                
                yield return request.SendWebRequest();
                
                if (request.result == UnityWebRequest.Result.Success)
                {
                    Debug.Log("[CulturalTranslate] Session ended");
                    sessionId = null;
                    OnDisconnected?.Invoke();
                }
                else
                {
                    Debug.LogError($"[CulturalTranslate] Error ending session: {request.error}");
                }
            }
        }
        
        void OnDestroy()
        {
            EndSession();
        }
    }
    
    // Data classes
    [Serializable]
    public class SessionRequest
    {
        public string source_language;
        public string target_language;
        public string source_culture;
        public string target_culture;
        public string type;
        public string cultural_adaptation_level;
    }
    
    [Serializable]
    public class SessionResponse
    {
        public string public_id;
        public string status;
    }
    
    [Serializable]
    public class TextTranslationRequest
    {
        public string session_id;
        public string text;
        public string source_language;
        public string target_language;
    }
    
    [Serializable]
    public class TranslationResponse
    {
        public string translated_text;
    }
    
    [Serializable]
    public class VoiceTranslationResponse
    {
        public string source_text;
        public string translated_text;
        public string audio_url;
    }
    
    [Serializable]
    public class TranslationResult
    {
        public string sourceText;
        public string translatedText;
        public string sourceLanguage;
        public string targetLanguage;
        public string audioUrl;
    }
}

File: public/sdk/cultural-translate-sdk.js
/**
 * Cultural Translate Gaming SDK
 * Version: 1.0.0
 * 
 * Real-time translation SDK for game developers
 * Supports voice and text translation with cultural adaptation
 */

class CulturalTranslateSDK {
    constructor(config) {
        this.apiKey = config.apiKey;
        this.gameId = config.gameId;
        this.baseUrl = config.baseUrl || 'https://culturaltranslate.com/api';
        this.wsUrl = config.wsUrl || 'wss://culturaltranslate.com:8080';
        
        this.session = null;
        this.ws = null;
        this.audioContext = null;
        this.mediaRecorder = null;
        this.isRecording = false;
        
        this.callbacks = {
            onTranslation: null,
            onError: null,
            onConnect: null,
            onDisconnect: null,
        };
        
        this.init();
    }
    
    /**
     * Initialize SDK
     */
    init() {
        console.log('[CulturalTranslate SDK] Initializing...');
        this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
    }
    
    /**
     * Create a new translation session
     */
    async createSession(options = {}) {
        try {
            const response = await fetch(`${this.baseUrl}/gaming/sessions`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.apiKey}`,
                    'X-Game-ID': this.gameId,
                },
                body: JSON.stringify({
                    source_language: options.sourceLanguage || 'en',
                    target_language: options.targetLanguage || 'ar',
                    source_culture: options.sourceCulture || 'en-US',
                    target_culture: options.targetCulture || 'ar-SA',
                    type: 'gaming',
                    cultural_adaptation_level: options.culturalLevel || 'standard',
                }),
            });
            
            if (!response.ok) {
                throw new Error(`Failed to create session: ${response.statusText}`);
            }
            
            this.session = await response.json();
            console.log('[CulturalTranslate SDK] Session created:', this.session.public_id);
            
            // Connect to WebSocket
            await this.connectWebSocket();
            
            return this.session;
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Error creating session:', error);
            this.triggerCallback('onError', error);
            throw error;
        }
    }
    
    /**
     * Connect to WebSocket for real-time communication
     */
    async connectWebSocket() {
        return new Promise((resolve, reject) => {
            this.ws = new WebSocket(`${this.wsUrl}/app/${this.session.public_id}`);
            
            this.ws.onopen = () => {
                console.log('[CulturalTranslate SDK] WebSocket connected');
                this.triggerCallback('onConnect');
                resolve();
            };
            
            this.ws.onmessage = (event) => {
                const data = JSON.parse(event.data);
                this.handleWebSocketMessage(data);
            };
            
            this.ws.onerror = (error) => {
                console.error('[CulturalTranslate SDK] WebSocket error:', error);
                this.triggerCallback('onError', error);
                reject(error);
            };
            
            this.ws.onclose = () => {
                console.log('[CulturalTranslate SDK] WebSocket disconnected');
                this.triggerCallback('onDisconnect');
            };
        });
    }
    
    /**
     * Handle WebSocket messages
     */
    handleWebSocketMessage(data) {
        switch (data.event) {
            case 'translation':
                this.triggerCallback('onTranslation', data.payload);
                break;
            case 'error':
                this.triggerCallback('onError', data.payload);
                break;
            default:
                console.log('[CulturalTranslate SDK] Unknown event:', data.event);
        }
    }
    
    /**
     * Translate text
     */
    async translateText(text, options = {}) {
        try {
            const response = await fetch(`${this.baseUrl}/gaming/translate/text`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${this.apiKey}`,
                    'X-Game-ID': this.gameId,
                },
                body: JSON.stringify({
                    session_id: this.session?.public_id,
                    text: text,
                    source_language: options.sourceLanguage || 'en',
                    target_language: options.targetLanguage || 'ar',
                }),
            });
            
            if (!response.ok) {
                throw new Error(`Translation failed: ${response.statusText}`);
            }
            
            const result = await response.json();
            return result.translated_text;
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Translation error:', error);
            this.triggerCallback('onError', error);
            throw error;
        }
    }
    
    /**
     * Start voice recording
     */
    async startVoiceRecording() {
        try {
            const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
            
            this.mediaRecorder = new MediaRecorder(stream);
            this.audioChunks = [];
            
            this.mediaRecorder.ondataavailable = (event) => {
                this.audioChunks.push(event.data);
            };
            
            this.mediaRecorder.onstop = async () => {
                const audioBlob = new Blob(this.audioChunks, { type: 'audio/webm' });
                await this.translateVoice(audioBlob);
                this.audioChunks = [];
            };
            
            this.mediaRecorder.start();
            this.isRecording = true;
            
            console.log('[CulturalTranslate SDK] Voice recording started');
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Error starting recording:', error);
            this.triggerCallback('onError', error);
            throw error;
        }
    }
    
    /**
     * Stop voice recording
     */
    stopVoiceRecording() {
        if (this.mediaRecorder && this.isRecording) {
            this.mediaRecorder.stop();
            this.isRecording = false;
            console.log('[CulturalTranslate SDK] Voice recording stopped');
        }
    }
    
    /**
     * Translate voice audio
     */
    async translateVoice(audioBlob) {
        try {
            const formData = new FormData();
            formData.append('audio', audioBlob, 'voice.webm');
            formData.append('session_id', this.session?.public_id);
            
            const response = await fetch(`${this.baseUrl}/gaming/translate/voice`, {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'X-Game-ID': this.gameId,
                },
                body: formData,
            });
            
            if (!response.ok) {
                throw new Error(`Voice translation failed: ${response.statusText}`);
            }
            
            const result = await response.json();
            this.triggerCallback('onTranslation', result);
            
            return result;
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Voice translation error:', error);
            this.triggerCallback('onError', error);
            throw error;
        }
    }
    
    /**
     * Play translated audio
     */
    async playTranslatedAudio(audioUrl) {
        try {
            const audio = new Audio(audioUrl);
            await audio.play();
        } catch (error) {
            console.error('[CulturalTranslate SDK] Error playing audio:', error);
            this.triggerCallback('onError', error);
        }
    }
    
    /**
     * Set callback functions
     */
    on(event, callback) {
        if (this.callbacks.hasOwnProperty(`on${event.charAt(0).toUpperCase() + event.slice(1)}`)) {
            this.callbacks[`on${event.charAt(0).toUpperCase() + event.slice(1)}`] = callback;
        }
    }
    
    /**
     * Trigger callback
     */
    triggerCallback(callbackName, data) {
        if (this.callbacks[callbackName] && typeof this.callbacks[callbackName] === 'function') {
            this.callbacks[callbackName](data);
        }
    }
    
    /**
     * End session
     */
    async endSession() {
        try {
            if (this.ws) {
                this.ws.close();
            }
            
            if (this.session) {
                await fetch(`${this.baseUrl}/gaming/sessions/${this.session.public_id}`, {
                    method: 'DELETE',
                    headers: {
                        'Authorization': `Bearer ${this.apiKey}`,
                        'X-Game-ID': this.gameId,
                    },
                });
            }
            
            this.session = null;
            console.log('[CulturalTranslate SDK] Session ended');
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Error ending session:', error);
            this.triggerCallback('onError', error);
        }
    }
    
    /**
     * Get session statistics
     */
    async getStats() {
        try {
            const response = await fetch(`${this.baseUrl}/gaming/sessions/${this.session.public_id}/stats`, {
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'X-Game-ID': this.gameId,
                },
            });
            
            if (!response.ok) {
                throw new Error(`Failed to get stats: ${response.statusText}`);
            }
            
            return await response.json();
            
        } catch (error) {
            console.error('[CulturalTranslate SDK] Error getting stats:', error);
            this.triggerCallback('onError', error);
            throw error;
        }
    }
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = CulturalTranslateSDK;
} else if (typeof define === 'function' && define.amd) {
    define([], function() {
        return CulturalTranslateSDK;
    });
} else {
    window.CulturalTranslateSDK = CulturalTranslateSDK;
}
